templates/smtp-credentials.yaml

```diff
@@ -0,0 +1,272 @@
+# MIT License https://github.com/tsamuel33/aws-smtp-credentials/blob/main/LICENSE
+# Template repository https://github.com/tsamuel33/aws-smtp-credentials
+
+AWSTemplateFormatVersion: 2010-09-09
+Description: >-
+  This template creates a custom SMTP resource that stores the SMTP
+  credentials in Parameter Store under the names "smtp-username" and
+  "smtp-password".
+
+Parameters:
+
+  IamAccessKeyVersion:
+    Type: Number
+    Default: 1
+    Description: >-
+      Version number of the AWS access keys.
+      Increment this number to rotate the keys.
+    MinValue: 1
+
+  TemplateRootUrl:
+    Type: String
+    Description: >-
+      URL of S3 bucket where templates used for nested stacks are deployed
+    ConstraintDescription: Must be a valid S3 HTTP URL
+
+Resources:
+
+  SmtpUserGroup:
+    Type: AWS::IAM::Group
+    Properties:
+      GroupName: SMTPUserGroup
+
+  SmtpUser:
+    Type: AWS::IAM::User
+    Properties:
+      UserName: SMTPUser
+      Groups:
+        - !Ref SmtpUserGroup
+
+  SmtpUserPolicy:
+    Type: AWS::IAM::Policy
+    Properties:
+      PolicyName: SMTPUserPolicy
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Action: ses:SendRawEmail
+            Resource: '*'
+      Groups:
+        - !Ref SmtpUserGroup
+
+  SmtpUserAccessKey:
+    Type: AWS::IAM::AccessKey
+    Properties:
+      Serial: !Ref IamAccessKeyVersion
+      Status: Active
+      UserName: !Ref SmtpUser
+
+  SmtpPassword:
+    Type: Custom::SmtpPassword
+    Properties:
+      ServiceToken: !GetAtt SmtpLambdaFunction.Arn
+      Key: !GetAtt SmtpUserAccessKey.SecretAccessKey
+      ParameterType: password
+
+  SmtpUsername:
+    Type: Custom::SmtpUsername
+    Properties:
+      ServiceToken: !GetAtt SmtpLambdaFunction.Arn
+      Key: !Ref SmtpUserAccessKey
+      ParameterType: username
+
+  SmtpLambdaExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      RoleName: SMTPLambdaExecutionRole
+      Description: Role assumed by Lambda to generate SMTP credentials
+      AssumeRolePolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service: lambda.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      Policies:
+        - PolicyName: InlineSMTPLambdaExecutionRolePolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+                - Effect: Allow
+                  Action:
+                    - logs:CreateLogStream
+                    - logs:PutLogEvents
+                  Resource:
+                    - !Sub "arn:${AWS::Partition}:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/SMTPCredentialsLambdaFunction"
+                    - !Sub "arn:${AWS::Partition}:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/SMTPCredentialsLambdaFunction:log-stream:*"
+                - Effect: Allow
+                  Action:
```

Comment:
it looks like the calculate_key() function in the lambda code is doing some decrypting to get the pwd, so don't you want the policy from the original code?
```
                - Effect: Allow
                  Action:
                    - kms:Decrypt
                    - kms:DescribeKey
                    - kms:CreateGrant
                    - kms:Encrypt
                  Resource:
                    - !Ref KmsKeyArn
```

Reply:
No, I added SmtpLambdaExecutionRole as a prinicipal on the SmtpKeyStack. The way he wrote it actually doesn't work -- there was a change in 2019.